File: pkg/parser/declarations.go
package parser

import (
	"github.com/bioflowy/flowy/pkg/expr"
	"github.com/bioflowy/flowy/pkg/tree"
)

// parseDeclaration parses any declaration according to WDL grammar:
// ?any_decl: unbound_decl | bound_decl
func (p *Parser) parseDeclaration() (*tree.Decl, bool) {
	pos := p.currentPosition()

	// Parse optional env modifier for task declarations
	// isEnv := false
	if p.currentTokenIs(TokenEnv) {
		// isEnv = true
		p.nextToken()
		// Note: We'll store this information but pkg/tree doesn't currently 
		// have a field for env declarations, so this is for future extension
	}

	// Parse type
	declType, ok := p.parseType()
	if !ok {
		return nil, false
	}

	// Parse identifier
	name, ok := p.parseIdentifier()
	if !ok {
		return nil, false
	}

	// Check for initialization expression
	var initExpr expr.Expr
	if p.currentTokenIs(TokenAssign) {
		p.nextToken() // consume '='
		
		// Parse initialization expression
		// Note: parseExpression will be implemented in expressions.go
		initExpr, ok = p.parseExpression()
		if !ok {
			// For now, create a placeholder
			p.addError(NewParseError(
				p.currentPosition(),
				"expression parsing not yet implemented",
				[]TokenType{},
				p.currentToken,
			))
			return nil, false
		}
	}

	return tree.NewDecl(name, declType, initExpr, pos), true
}

// parseUnboundDeclaration parses an unbound declaration:
// unbound_decl: type CNAME -> decl
func (p *Parser) parseUnboundDeclaration() (*tree.Decl, bool) {
	pos := p.currentPosition()

	// Parse type
	declType, ok := p.parseType()
	if !ok {
		return nil, false
	}

	// Parse identifier
	name, ok := p.parseIdentifier()
	if !ok {
		return nil, false
	}

	// Unbound declarations should not have assignments
	if p.currentTokenIs(TokenAssign) {
		p.addError(NewParseError(
			p.currentPosition(),
			"unbound declaration cannot have initialization",
			[]TokenType{},
			p.currentToken,
		))
		return nil, false
	}

	// No initialization for unbound declarations
	return tree.NewDecl(name, declType, nil, pos), true
}

// parseBoundDeclaration parses a bound declaration:
// bound_decl: type CNAME "=" expr -> decl
func (p *Parser) parseBoundDeclaration() (*tree.Decl, bool) {
	pos := p.currentPosition()

	// Parse type
	declType, ok := p.parseType()
	if !ok {
		return nil, false
	}

	// Parse identifier
	name, ok := p.parseIdentifier()
	if !ok {
		return nil, false
	}

	// Expect assignment
	if !p.consume(TokenAssign) {
		return nil, false
	}

	// Parse initialization expression
	initExpr, ok := p.parseExpression()
	if !ok {
		return nil, false
	}

	return tree.NewDecl(name, declType, initExpr, pos), true
}

// parseInputDeclarations parses input declarations:
// input_decls: "input" "{" any_decl* "}"
func (p *Parser) parseInputDeclarations() (*tree.Input, bool) {
	pos := p.currentPosition()

	if !p.consume(TokenInput) {
		return nil, false
	}

	var declarations []*tree.Decl
	ok := p.parseBlock(func() bool {
		for !p.currentTokenIs(TokenRightBrace) && !p.IsAtEnd() {
			if decl, ok := p.parseDeclaration(); ok {
				declarations = append(declarations, decl)
			} else {
				// Try to recover
				p.synchronize()
				break
			}
			
			p.skipCommentsAndNewlines()
		}
		return true
	})

	if !ok {
		return nil, false
	}

	return tree.NewInput(declarations, pos), true
}

// parseTaskInputDeclarations parses task input declarations with optional env:
// task_input_decls: "input" "{" task_env_decl* "}" -> input_decls
// task_env_decl: ENV? any_decl
func (p *Parser) parseTaskInputDeclarations() (*tree.Input, bool) {
	// For now, this is the same as regular input declarations
	// The env modifier is parsed but not yet stored in the tree structure
	return p.parseInputDeclarations()
}

// parseOutputDeclarations parses output declarations:
// output_decls: "output" "{" bound_decl* "}"
func (p *Parser) parseOutputDeclarations() (*tree.Output, bool) {
	pos := p.currentPosition()

	if !p.consume(TokenOutput) {
		return nil, false
	}

	var declarations []*tree.Decl
	ok := p.parseBlock(func() bool {
		for !p.currentTokenIs(TokenRightBrace) && !p.IsAtEnd() {
			// Output declarations must be bound (have initialization)
			if decl, ok := p.parseBoundDeclaration(); ok {
				declarations = append(declarations, decl)
			} else {
				// Try to recover
				p.synchronize()
				break
			}
			
			p.skipCommentsAndNewlines()
		}
		return true
	})

	if !ok {
		return nil, false
	}

	return tree.NewOutput(declarations, pos), true
}

// parseStruct parses a struct definition:
// struct: "struct" CNAME "{" unbound_decl* "}"
func (p *Parser) parseStruct() (*tree.StructTypeDef, bool) {
	pos := p.currentPosition()

	if !p.consume(TokenStruct) {
		return nil, false
	}

	// Parse struct name
	name, ok := p.parseIdentifier()
	if !ok {
		return nil, false
	}

	var members []*tree.StructMember
	ok = p.parseBlock(func() bool {
		for !p.currentTokenIs(TokenRightBrace) && !p.IsAtEnd() {
			// Parse struct member (unbound declaration)
			if member, ok := p.parseStructMember(); ok {
				members = append(members, member)
			} else {
				// Try to recover
				p.synchronize()
				break
			}
			
			p.skipCommentsAndNewlines()
		}
		return true
	})

	if !ok {
		return nil, false
	}

	return tree.NewStructTypeDef(name, members, pos), true
}

// parseStructMember parses a struct member (unbound declaration)
func (p *Parser) parseStructMember() (*tree.StructMember, bool) {
	pos := p.currentPosition()

	// Parse type
	memberType, ok := p.parseType()
	if !ok {
		return nil, false
	}

	// Parse member name
	name, ok := p.parseIdentifier()
	if !ok {
		return nil, false
	}

	return tree.NewStructMember(name, memberType, pos), true
}

// parseDeclarationList parses a list of declarations
func (p *Parser) parseDeclarationList(terminator TokenType) ([]*tree.Decl, bool) {
	var declarations []*tree.Decl

	for !p.currentTokenIs(terminator) && !p.IsAtEnd() {
		if decl, ok := p.parseDeclaration(); ok {
			declarations = append(declarations, decl)
		} else {
			return nil, false
		}
		
		p.skipCommentsAndNewlines()
	}

	return declarations, true
}

// parseUnboundDeclarationList parses a list of unbound declarations
func (p *Parser) parseUnboundDeclarationList(terminator TokenType) ([]*tree.Decl, bool) {
	var declarations []*tree.Decl

	for !p.currentTokenIs(terminator) && !p.IsAtEnd() {
		if decl, ok := p.parseUnboundDeclaration(); ok {
			declarations = append(declarations, decl)
		} else {
			return nil, false
		}
		
		p.skipCommentsAndNewlines()
	}

	return declarations, true
}

// parseBoundDeclarationList parses a list of bound declarations
func (p *Parser) parseBoundDeclarationList(terminator TokenType) ([]*tree.Decl, bool) {
	var declarations []*tree.Decl

	for !p.currentTokenIs(terminator) && !p.IsAtEnd() {
		if decl, ok := p.parseBoundDeclaration(); ok {
			declarations = append(declarations, decl)
		} else {
			return nil, false
		}
		
		p.skipCommentsAndNewlines()
	}

	return declarations, true
}

// parseStructMemberList parses a list of struct members
func (p *Parser) parseStructMemberList(terminator TokenType) ([]*tree.StructMember, bool) {
	var members []*tree.StructMember

	for !p.currentTokenIs(terminator) && !p.IsAtEnd() {
		if member, ok := p.parseStructMember(); ok {
			members = append(members, member)
		} else {
			return nil, false
		}
		
		p.skipCommentsAndNewlines()
	}

	return members, true
}

// isDeclarationStart returns true if current token can start a declaration
func (p *Parser) isDeclarationStart() bool {
	// Declarations start with types or env keyword
	return p.isTypeStart() || p.currentTokenIs(TokenEnv)
}

// parseVariableDeclaration parses a variable declaration in various contexts
func (p *Parser) parseVariableDeclaration() (*tree.Decl, bool) {
	return p.parseDeclaration()
}

// validateDeclarationName validates that a declaration name is valid
func (p *Parser) validateDeclarationName(name string) bool {
	// Check if it's a valid identifier
	if !isValidIdentifier(name) {
		return false
	}

	// Check if it conflicts with reserved keywords
	switch name {
	case "true", "false", "None":
		return false // These are literal values
	case "version", "import", "as", "alias", "workflow", "task", 
		 "scatter", "if", "then", "else", "call", "struct":
		return false // These are core language keywords that cannot be variable names
	default:
		return true
	}
}

File: src/parser/declarations.rs
//! Token-based declaration parsing for WDL

use super::token_stream::TokenStream;
use super::tokens::Token;
use super::parser_utils::ParseResult;
use super::types::parse_type;
use super::expressions::parse_expression;
use crate::tree::Declaration;
use crate::error::WdlError;

/// Parse a declaration: Type name = expression
pub fn parse_declaration(stream: &mut TokenStream, id_prefix: &str) -> ParseResult<Declaration> {
    let pos = stream.current_position();
    
    // Parse the type
    let decl_type = parse_type(stream)?;
    
    // Parse the variable name
    let name = match stream.peek_token() {
        Some(Token::Identifier(n)) => {
            let name = n.clone();
            stream.next();
            name
        }
        _ => {
            return Err(WdlError::syntax_error(
                stream.current_position(),
                "Expected variable name after type".to_string(),
                "1.0".to_string(),
                None,
            ));
        }
    };
    
    // Check for assignment
    let expr = if stream.peek_token() == Some(&Token::Assign) {
        stream.next(); // consume =
        Some(parse_expression(stream)?)
    } else {
        None
    };
    
    Ok(Declaration::new(
        pos,
        decl_type,
        name,
        expr,
        id_prefix,
    ))
}

/// Parse an input declaration (may have no initializer in inputs section)
pub fn parse_input_declaration(stream: &mut TokenStream) -> ParseResult<Declaration> {
    parse_declaration(stream, "input")
}

/// Parse an output declaration (usually has an initializer)
pub fn parse_output_declaration(stream: &mut TokenStream) -> ParseResult<Declaration> {
    parse_declaration(stream, "output")
}

/// Parse a list of declarations
pub fn parse_declaration_list(stream: &mut TokenStream, id_prefix: &str) -> ParseResult<Vec<Declaration>> {
    let mut declarations = Vec::new();
    
    // Parse declarations until we hit a closing brace or EOF
    while stream.peek_token() != Some(&Token::RightBrace) && !stream.is_eof() {
        // Skip any newlines
        while stream.peek_token() == Some(&Token::Newline) {
            stream.next();
        }
        
        // Check if we've reached the end
        if stream.peek_token() == Some(&Token::RightBrace) || stream.is_eof() {
            break;
        }
        
        // Parse a declaration
        let decl = parse_declaration(stream, id_prefix)?;
        declarations.push(decl);
        
        // Consume optional newlines after declaration
        while stream.peek_token() == Some(&Token::Newline) {
            stream.next();
        }
    }
    
    Ok(declarations)
}

/// Parse an input section: input { declarations }
pub fn parse_input_section(stream: &mut TokenStream) -> ParseResult<Vec<Declaration>> {
    // Expect "input" keyword
    match stream.peek_token() {
        Some(Token::Keyword(kw)) if kw == "input" => {
            stream.next();
        }
        _ => {
            return Err(WdlError::syntax_error(
                stream.current_position(),
                "Expected 'input' keyword".to_string(),
                "1.0".to_string(),
                None,
            ));
        }
    }
    
    // Expect opening brace
    stream.expect(Token::LeftBrace)?;
    
    // Parse declarations
    let declarations = parse_declaration_list(stream, "input")?;
    
    // Expect closing brace
    stream.expect(Token::RightBrace)?;
    
    Ok(declarations)
}

/// Parse an output section: output { declarations }
pub fn parse_output_section(stream: &mut TokenStream) -> ParseResult<Vec<Declaration>> {
    // Expect "output" keyword
    match stream.peek_token() {
        Some(Token::Keyword(kw)) if kw == "output" => {
            stream.next();
        }
        _ => {
            return Err(WdlError::syntax_error(
                stream.current_position(),
                "Expected 'output' keyword".to_string(),
                "1.0".to_string(),
                None,
            ));
        }
    }
    
    // Expect opening brace
    stream.expect(Token::LeftBrace)?;
    
    // Parse declarations
    let declarations = parse_declaration_list(stream, "output")?;
    
    // Expect closing brace
    stream.expect(Token::RightBrace)?;
    
    Ok(declarations)
}

/// Parse a private declarations section (for workflow/task body)
pub fn parse_private_declaration(stream: &mut TokenStream) -> ParseResult<Declaration> {
    parse_declaration(stream, "decl")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::token_stream::TokenStream;
    use crate::types::Type;
    
    #[test]
    fn test_parse_simple_declaration() {
        let mut stream = TokenStream::new("Int x = 42", "1.0").unwrap();
        let result = parse_declaration(&mut stream, "test");
        assert!(result.is_ok());
        
        let decl = result.unwrap();
        assert_eq!(decl.name, "x");
        assert_eq!(decl.decl_type, Type::int(false));
        assert!(decl.expr.is_some());
    }
    
    #[test]
    fn test_parse_declaration_without_init() {
        let mut stream = TokenStream::new("String name", "1.0").unwrap();
        let result = parse_declaration(&mut stream, "test");
        assert!(result.is_ok());
        
        let decl = result.unwrap();
        assert_eq!(decl.name, "name");
        assert_eq!(decl.decl_type, Type::string(false));
        assert!(decl.expr.is_none());
    }
    
    #[test]
    fn test_parse_optional_declaration() {
        let mut stream = TokenStream::new("File? optional_file", "1.0").unwrap();
        let result = parse_declaration(&mut stream, "test");
        assert!(result.is_ok());
        
        let decl = result.unwrap();
        assert_eq!(decl.name, "optional_file");
        assert_eq!(decl.decl_type, Type::file(true));
        assert!(decl.expr.is_none());
    }
    
    #[test]
    fn test_parse_array_declaration() {
        let mut stream = TokenStream::new("Array[String] items = [\"a\", \"b\"]", "1.0").unwrap();
        let result = parse_declaration(&mut stream, "test");
        assert!(result.is_ok());
        
        let decl = result.unwrap();
        assert_eq!(decl.name, "items");
        assert!(matches!(decl.decl_type, Type::Array { .. }));
        assert!(decl.expr.is_some());
    }
    
    #[test]
    fn test_parse_input_section() {
        let input = r#"input {
            String name
            Int count = 10
            File? optional_file
        }"#;
        
        let mut stream = TokenStream::new(input, "1.0").unwrap();
        let result = parse_input_section(&mut stream);
        assert!(result.is_ok());
        
        let declarations = result.unwrap();
        assert_eq!(declarations.len(), 3);
        assert_eq!(declarations[0].name, "name");
        assert_eq!(declarations[1].name, "count");
        assert_eq!(declarations[2].name, "optional_file");
    }
    
    #[test]
    fn test_parse_output_section() {
        let input = r#"output {
            String result = "done"
            Array[File] files = glob("*.txt")
        }"#;
        
        let mut stream = TokenStream::new(input, "1.0").unwrap();
        let result = parse_output_section(&mut stream);
        assert!(result.is_ok());
        
        let declarations = result.unwrap();
        assert_eq!(declarations.len(), 2);
        assert_eq!(declarations[0].name, "result");
        assert_eq!(declarations[1].name, "files");
    }
    
    #[test]
    fn test_parse_map_declaration() {
        let mut stream = TokenStream::new("Map[String, Int] counts = {\"a\": 1}", "1.0").unwrap();
        let result = parse_declaration(&mut stream, "test");
        assert!(result.is_ok());
        
        let decl = result.unwrap();
        assert_eq!(decl.name, "counts");
        assert!(matches!(decl.decl_type, Type::Map { .. }));
        assert!(decl.expr.is_some());
    }
    
    #[test]
    fn test_parse_complex_declaration() {
        let mut stream = TokenStream::new(
            "Array[Pair[String, File]]? results = [(\"a\", \"file.txt\")]",
            "1.0"
        ).unwrap();
        let result = parse_declaration(&mut stream, "test");
        assert!(result.is_ok());
        
        let decl = result.unwrap();
        assert_eq!(decl.name, "results");
        // Check it's an optional array of pairs
        if let Type::Array { optional, .. } = decl.decl_type {
            assert!(optional);
        } else {
            panic!("Expected Array type");
        }
    }
}